pkg/kubelet/userns_manager.go

```diff
@@ -0,0 +1,503 @@
+	if err = json.Unmarshal([]byte(content), &userNs); err != nil {
+		err = fmt.Errorf("can't parse file: %w", err)
+		return
+	}
+
+	if len(userNs.UIDMappings) != 1 {
+		err = fmt.Errorf("invalid user namespace configuration: no more than one mapping allowed.")
+		return
+	}
+
+	if len(userNs.UIDMappings) != len(userNs.GIDMappings) {
+		err = fmt.Errorf("invalid user namespace configuration: GID and UID mappings should be identical.")
+		return
+	}
+
+	if userNs.UIDMappings[0] != userNs.GIDMappings[0] {
+		err = fmt.Errorf("invalid user namespace configuration: GID and UID mapping should be identical")
+		return
+	}
+
+	// We don't produce configs without root mapped and some runtimes assume it is mapped.
+	// Validate the file has something we produced and can digest.
+	if userNs.UIDMappings[0].ContainerId != 0 {
+		err = fmt.Errorf("invalid user namespace configuration: UID 0 must be mapped")
+		return
+	}
+
+	if userNs.GIDMappings[0].ContainerId != 0 {
+		err = fmt.Errorf("invalid user namespace configuration: GID 0 must be mapped")
+		return
+	}
+
+	hostId := userNs.UIDMappings[0].HostId
+	length := userNs.UIDMappings[0].Length
+
+	err = m.record(pod, hostId, length)
+	return
+}
+
+func (m *usernsManager) createUserNs(pod *v1.Pod) (userNs userNamespace, err error) {
+	firstID, length, err := m.allocateOne(string(pod.UID))
+	if err != nil {
+		return
+	}
+
+	defer func() {
+		if err != nil {
+			m.Release(string(pod.UID))
+		}
+	}()
+
+	userNs = userNamespace{
+		UIDMappings: []idMapping{
+			{
+				ContainerId: 0,
+				HostId:      firstID,
+				Length:      length,
+			},
+		},
+		GIDMappings: []idMapping{
+			{
+				ContainerId: 0,
+				HostId:      firstID,
+				Length:      length,
+			},
+		},
+	}
+
+	return userNs, m.writeMappingsToFile(pod.UID, userNs)
+}
+
+// GetUserNamespaceMappings returns the configuration for the sandbox user namespace
+func (m *usernsManager) GetUserNamespaceMappings(pod *v1.Pod) (*runtimeapi.UserNamespace, error) {
+	if !utilfeature.DefaultFeatureGate.Enabled(features.UserNamespacesSupport) {
+		return nil, nil
+	}
+
+	m.Lock()
+	defer m.Unlock()
+
+	if pod.Spec.HostUsers == nil || *pod.Spec.HostUsers == true {
+		return &runtimeapi.UserNamespace{
+			Mode: runtimeapi.NamespaceMode_NODE,
+		}, nil
+	}
+
+	content, err := m.readMappingsFromFile(pod.UID)
+	if err != nil && err != utilstore.ErrKeyNotFound {
+		return nil, err
+	}
+
+	var userNs userNamespace
+	if string(content) != "" {
+		userNs, err = m.parseUserNsFileAndRecord(pod.UID, content)
+		if err != nil {
+			return nil, err
+		}
+	} else {
+		userNs, err = m.createUserNs(pod)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	// A new pod with userns is being created.
+	if m.currPods >= maxPods {
+		return nil, fmt.Errorf("limit on count of pods with user namespaces exceeded (limit is %v)", maxPods)
+	}
+	m.currPods++
+
+	var uids []*runtimeapi.IDMapping
+	var gids []*runtimeapi.IDMapping
+
+	for _, u := range userNs.UIDMappings {
```

Comment:
if we only allow one UIDMapping/GIDMapping, then this for loop is not needed right?

Reply:
I thought it would improve readability and be future-proof since we'll need to drop the limitation only in one place (`GetUserNamespaceMappings` doesn't need to know what happens in `createUserNs`).

If you prefer I can drop the loop.

Reply:
oh i am fine with the for loop here actually. I just saw the other comment [1] indicating otherwise, so checking if we want to keep consistent. 


[1] https://github.com/kubernetes/kubernetes/pull/111090/files#r919940061